google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/TopicAdminSettings.java

```diff
@@ -526,147 +179,54 @@ protected Builder() {
-
-      publishSettings =
-          BatchingCallSettings.newBuilder(PUBLISH_BATCHING_DESC)
-              .setBatchingSettings(BatchingSettings.newBuilder().build());
-
-      getTopicSettings = UnaryCallSettings.newUnaryCallSettingsBuilder();
-
-      listTopicsSettings = PagedCallSettings.newBuilder(LIST_TOPICS_PAGE_STR_FACT);
-
-      listTopicSubscriptionsSettings =
-          PagedCallSettings.newBuilder(LIST_TOPIC_SUBSCRIPTIONS_PAGE_STR_FACT);
-
-      deleteTopicSettings = UnaryCallSettings.newUnaryCallSettingsBuilder();
-
-      setIamPolicySettings = UnaryCallSettings.newUnaryCallSettingsBuilder();
-
-      getIamPolicySettings = UnaryCallSettings.newUnaryCallSettingsBuilder();
+      stubBuilder = PublisherStubSettings.newBuilder(clientContext);
+    }
 
-      testIamPermissionsSettings = UnaryCallSettings.newUnaryCallSettingsBuilder();
+    private static Builder createDefault() {
+      Builder builder = new Builder((ClientContext) null);
+      builder.setTransportChannelProvider(PublisherStubSettings.defaultTransportChannelProvider());
+      builder.setCredentialsProvider(PublisherStubSettings.defaultCredentialsProviderBuilder().build());
+      builder.setInternalHeaderProvider(PublisherStubSettings.defaultApiClientHeaderProviderBuilder().build());
+      builder.setEndpoint(PublisherStubSettings.getDefaultEndpoint());
+      builder.stubBuilder = PublisherStubSettings.newBuilder();
+      return builder;
+    }
 
-      unaryMethodSettingsBuilders =
-          ImmutableList.<UnaryCallSettings.Builder<?, ?>>of(
-              createTopicSettings,
-              updateTopicSettings,
-              publishSettings,
-              getTopicSettings,
-              listTopicsSettings,
-              listTopicSubscriptionsSettings,
-              deleteTopicSettings,
-              setIamPolicySettings,
-              getIamPolicySettings,
-              testIamPermissionsSettings);
+    protected Builder(TopicAdminSettings settings) {
+      super(settings.getStubSettings());
+      stubBuilder = settings.getStubSettings().toBuilder();
+    }
 
-      initDefaults(this);
+    protected Builder(PublisherStubSettings stubSettings) {
+      super(stubSettings);
+      stubBuilder = stubSettings.toBuilder();
     }
 
-    private static Builder createDefault() {
-      Builder builder = new Builder((ClientContext) null);
-      builder.setTransportChannelProvider(defaultTransportChannelProvider());
-      builder.setCredentialsProvider(defaultCredentialsProviderBuilder().build());
-      builder.setInternalHeaderProvider(defaultApiClientHeaderProviderBuilder().build());
-      builder.setEndpoint(getDefaultEndpoint());
-      return initDefaults(builder);
+    @Override
+    public Builder setTransportChannelProvider(TransportChannelProvider channelProvider) {
+      super.setTransportChannelProvider(channelProvider);
+      stubBuilder.setTransportChannelProvider(channelProvider);
+      return this;
```

Comment:
This makes a confusing situation where it's unclear which object holds the providers that will be used at client construction time. Also, I want to avoid generating general methods at the API-specific level. This probably means we need to support composition at the Gax level itself: we should probably have the ClientSettings of the TopicAdminSettings object hold the PublisherStubSettings object instead of TopicAdminSettings holding it.

Reply:
That is a problem for sure. So the ClientSettings should contain another ClientSettings field? That seems confusing too.

Reply:
Ok I think the rabbit hole is going slightly deeper - it appears that we need to create a StubSettings class and move the ClientSettings implementation into it, while leaving the ClientSettings surface itself intact. (This is just a parallel change to the API-specific classes.)

Reply:
lol the things we do for abstraction
gax-java PR at https://github.com/googleapis/gax-java/pull/459

still updating this PR to reflect gax changes...